sklearn/preprocessing/discretization.py

```diff
@@ -0,0 +1,341 @@
+        -------
+        self
+        """
+
+        if self.n_bins < 2:
+            raise ValueError("Discretizer received an invalid number "
+                             "of bins. Received {0}, expected at least 2"
+                             .format(self.n_bins))
+
+        X = check_array(X, accept_sparse=self.sparse_formats)
+        self.n_features_ = X.shape[1]
+
+        # Set the indices of continuous features
+        self._set_continuous_features()
+
+        continuous = X[:, self.continuous_features_]
+
+        if sp.issparse(X):
+            self.min_ = csr_csc_min_axis0(continuous)
+            self.max_ = csr_csc_max_axis0(continuous)
+        else:
+            self.min_ = continuous.min(axis=0)
+            self.max_ = continuous.max(axis=0)
+
+        searched_points = list()
+        zero_intervals = list()
+
+        for min_, max_ in zip(self.min_, self.max_):
+
+            points = np.linspace(min_, max_, num=self.n_bins, endpoint=False)[1:]
+
+            # Get index of where zero goes. Omit this index in
+            # the rebuilt array
+            # TODO: Watch out for when there is only two intervals
+            zero_index = np.searchsorted(points, 0, side='right')
+
+            # Case when all values are positive
+            if zero_index == 0:
+                zero_int = (-np.inf, points[zero_index])
+                searched = points[1:]
+
+            # Case when all values are negative
+            elif zero_index == len(points):
+                zero_int = (points[zero_index - 1], np.inf)
+                searched = points[:-1]
+
+            else:
+                zero_int = (points[zero_index - 1], points[zero_index])
+                searched = np.hstack((points[:zero_index - 1], points[zero_index:]))
+            zero_intervals.append(zero_int)
+
+            searched_points.append(searched.reshape(-1, 1))
+
+        self.searched_points_ = np.hstack(searched_points)
+        self.zero_intervals_ = zero_intervals
+        return self
+
+    @property
+    def n_continuous_features_(self):
+        if not self.continuous_features_:
+            return None
+        return len(self.continuous_features_)
+
+    @property
+    def cut_points_(self):
+        if not self.zero_intervals_:
+            return None
+        zero_intervals = self.zero_intervals_
+        searched_points = self.searched_points_
+
+        cut_points = list()
+        for (lower, upper), col in zip(zero_intervals, searched_points.T):
+
+            if lower == -np.inf:
+                cut_column = np.insert(col, 0, upper)
+            elif upper == np.inf:
+                cut_column = np.append(col, lower)
+            else:
+                zero_index = np.searchsorted(col, lower)
+                cut_column = np.insert(col, zero_index, lower)
+            cut_points.append(cut_column.reshape(-1, 1))
+        cut_points = np.hstack(cut_points)
+        return cut_points
+
+    @property
+    def cat_features_(self):
```

Comment:
Hmm, do you think this function should be renamed as `categorical_features_`? I guess `cat_features` is an abbreviation, which might not be very ideal.